packages/assets-controllers/src/NftController.ts

```diff
@@ -266,24 +456,25 @@ export class NftController extends BaseControllerV1<NftConfig, NftState> {
    * @param passedConfig.userAddress - the address passed through the NFT detection flow to ensure assets are stored to the correct account
    * @param passedConfig.chainId - the chainId passed through the NFT detection flow to ensure assets are stored to the correct account
    */
-  private updateNestedNftState(
-    newCollection: Nft[] | NftContract[],
-    baseStateKey: 'allNfts' | 'allNftContracts',
-    { userAddress, chainId }: { userAddress: string; chainId: Hex },
+  #updateNestedNftState<
+    T extends 'allNfts' | 'allNftContracts',
+    U extends T extends 'allNfts' ? Nft[] : NftContract[],
+  >(
+    newCollection: U,
+    baseStateKey: T,
+    { userAddress, chainId }: { userAddress: string; chainId: string },
   ) {
```

Comment:
Could you explain the new generic params for me? Not sure what they're achieving here. 
- Widening `chainId` type from `Hex` to `string` seems like a loss in type safety.
- Isn't `T` is a string type while `U` is an array? Not sure how `U extends T` is possible.

Nit:
- We should use descriptive generic param names.
- Maybe we should consistently use the constant variable names for `allNfts` and `allNftContracts`

```suggestion
    Key extends 'allNfts' | 'allNftContracts',
    NftCollection extends Key extends 'allNfts' ? Nft[] : NftContract[],
  >(
    newCollection: NftCollection,
    baseStateKey: Key,
    { userAddress, chainId }: { userAddress: string; chainId: string },
```

Reply:
@MajorLift I am not sure what you mean by `Maybe we should consistently use the constant variable names for 'allNfts' and 'allNftContracts'
`

Whereas to answer the generics part:
This was done because this is the state type:

`export type NftControllerState = {
  allNftContracts: {
    [key: string]: {
      [chainId: Hex]: NftContract[];
    };
  };
  allNfts: {
    [key: string]: {
      [chainId: Hex]: Nft[];
    };
  };
  ignoredNfts: Nft[];
};
`
The function #updateNestedNftState uses two generic parameters: T and U.

1. Generic Parameter T:
	•	T represents the possible keys for the state properties, which are 'allNfts' or 'allNftContracts'.
	•	It helps in making sure that the function is type-safe and only allows valid keys to be passed as the baseStateKey.
2. Generic Parameter U:
	•	U is a conditional type that depends on the value of T.
	•	If T is 'allNfts', then U will be Nft[].
	•	If T is 'allNftContracts', then U will be NftContract[].
	•	This ensures that the newCollection parameter is of the correct type based on the value of baseStateKey.

Reply:
Very nice! I didn't know `extends` operator precedence worked like that. Thanks for the explanation. 

Looks like the conditional generic params are validating the inferred types of the input params. I'll have to see if I can apply this to https://github.com/MetaMask/core/issues/4213.

About the constant variable names, I was suggesting we use `typeof ALL_NFTS_STATE_KEY | typeof ALL_NFTS_CONTRACTS_STATE_KEY` over `allNfts | allNftContracts`. This is a nit though.